lib/logstorage: add `coalesce` pipe, update documentation

warkadiusz opened the pull request:
### Describe Your Changes

Implements the `coalesce` pipe for LogsQL, which returns the first non-empty value from a list of fields.

- Added `coalesce` implementation and tests
- Updated LogsQL documentation and examples

Fixes #690 

Example usage:

- Return new field
   ```logsql
   * | coalesce (user_id, username, email) as user
   ```
- Return new field with default:
   ```logsql
   * | coalesce (user_id, username, email) default "anonymous" as user
   ```
- Overwrite existing field with the first non-empty value:
   ```logsql
   * | coalesce (user_id, username, email) as user_id
   ```

Looking forward to your feedback and happy to make any changes needed.

### Checklist

The following checks are **mandatory**:

- [x] My change adheres to [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/victoriametrics/contributing/#pull-request-checklist).
- [x] My change adheres to [VictoriaMetrics development goals](https://docs.victoriametrics.com/victoriametrics/goals/).

warkadiusz committed:
lib/logstorage: add `coalesce` pipe, update documentation

`coalesce` pipe returns the first non-empty value from the specified list of fields in order, writing the result as `result_field`.
If all source fields are empty, the optional `default` value is used.

Fixes https://github.com/VictoriaMetrics/VictoriaLogs/issues/690

warkadiusz committed:
lib/logstorage: update `as` and `default` syntax order in `coalesce` pipe

Updates `default` and `as` order in the `coalesce` pipe parsing to better fit LogQL semantics,
as requested in https://github.com/VictoriaMetrics/VictoriaLogs/issues/690#issuecomment-4205764878

warkadiusz committed:
lib/logstorage: update documentation to reflect as/default order changes

Adds on d372c828544761571fc2fef67f9849572e5450c3, fixes syntax order in
the documentation.

func25 commented on lib/logstorage/pipe_coalesce.go:
```diff
@@ -0,0 +1,206 @@
+		logger.Panicf("BUG: pipeCoalesce must contain at least one srcField")
+	}
+
+	s := "coalesce(" + strings.Join(pc.srcFields, ", ") + ")"
```
`srcFields` are joined as-is, so field names that require quoting will produce an invalid query string

warkadiusz replied:
Fixed in a40a73bdc2ec0163b1c5ba3d128f4132b0d7cdb2 by processing `srcFields` with `quoteTokenIfNeeded()` before joining them

func25 commented on lib/logstorage/pipe_coalesce.go:
```diff
@@ -0,0 +1,206 @@
+
+func (pc *pipeCoalesce) isFixedOutputFieldsOrder() bool {
+	return false
+}
+
+func (pc *pipeCoalesce) updateNeededFields(pf *prefixfilter.Filter) {
+	if pf.MatchString(pc.dstField) {
+		pf.AddAllowFilters(pc.srcFields)
+	}
+
+	pf.AddDenyFilter(pc.dstField)
+}
```
`updateNeededFields()` looks wrong for the overwrite case e.g., `coalesce(a, b) as a`

warkadiusz replied:
Thanks, fixed.

func25 commented on lib/logstorage/pipe_coalesce.go:
```diff
@@ -0,0 +1,206 @@
+
+	shard.rc.name = pc.dstField
+
+	srcColumns := make([]*blockResultColumn, len(pc.srcFields))
```
It is better to move `srcColumns` to `pipeCoalesceProcessorShard` and to reuse it across `writeBlock()` calls.

warkadiusz replied:
Fixed in a40a73bdc2ec0163b1c5ba3d128f4132b0d7cdb2 too – moved `srcColumns` to `pipeCoalesceProcessorShard`. Thanks

func25 commented on lib/logstorage/pipe_coalesce_test.go:
```diff
@@ -0,0 +1,188 @@
+	})
+}
+
+func TestPipeCoalesceUpdateNeededFields(t *testing.T) {
+	f := func(s, allowFilters, denyFilters, allowFiltersExpected, denyFiltersExpected string) {
+		t.Helper()
+		expectPipeNeededFields(t, s, allowFilters, denyFilters, allowFiltersExpected, denyFiltersExpected)
+	}
```
Please add a case such as `coalesce(s1, s2) as s1` here

warkadiusz replied:
Thanks! Added more cases, also with different fields order.

func25 commented on docs/victorialogs/CHANGELOG.md:
```diff
@@ -22,6 +22,8 @@ according to the following docs:
 
 ## tip
 
+* FEATURE: [LogsQL](https://docs.victoriametrics.com/victorialogs/logsql/): add [`coalesce` pipe](https://docs.victoriametrics.com/victorialogs/logsql/#coalesce-pipe) for returning the first non-empty value from the specified list of fields, or a default value if all are empty. See [#690](https://github.com/VictoriaMetrics/VictoriaLogs/issues/690)
```
```suggestion
* FEATURE: [LogsQL](https://docs.victoriametrics.com/victorialogs/logsql/): add [`coalesce` pipe](https://docs.victoriametrics.com/victorialogs/logsql/#coalesce-pipe) for returning the first non-empty value from the given fields. It can also return a default value when all the fields are empty. See [#690](https://github.com/VictoriaMetrics/VictoriaLogs/issues/690).
```

warkadiusz replied:
Aslo fixed, thank you.

func25 reviewed (commented):
Looks good overall. Please fix the conflicts and resolve the comments

func25 commented:
@warkadiusz, do you want to continue maintaining this PR?

warkadiusz commented:
> @warkadiusz, do you want to continue maintaining this PR?

Hi, yes. Sorry for the delay. I will provide fixes this weekend, is this fine?

func25 commented:
Sounds good, thanks for confirming!

warkadiusz committed:
lib/logstorage: fix PR feedback, reported bugs

Fixes `srcFields` joining in `String()`, `updateNeededFields()`
implementation for overwrite cases, implements reuse of `srcColumns` in
`writeBlock`, adds new test cases for pipe_coalesce.

warkadiusz commented on lib/logstorage/pipe_coalesce_test.go:
```diff
@@ -0,0 +1,199 @@
+	f(`coalesce(foo, bar) default " " as result`)
+	f(`coalesce(foo, bar) default foobar as result`)
+	f(`coalesce(foo, bar) default "coalesce" as result`)
+	f(`coalesce("foo bar", "foo-bar") as result`)
```
Additional case for the parsing mistake that was pointed out in the review

warkadiusz commented:
@func25 thank you for the review. Conflicts resolved, issues fixed. Let me know if there's anything else that needs improvement before merging.

func25 reviewed (approved):
LGTM!

func25 committed:
Merge branch 'master' into add-coalesce-pipe

Signed-off-by: Phuong Le <39565248+func25@users.noreply.github.com>
